### PR TITLE
Base version on 'git describe --tags'

### PR DESCRIPTION
**Pull request by VannTen**
This makes the version based on `git describe --tags`, which is
basically `<most recent tag>-<number of commits since that tag>-<commit
hash>`

This is useful for us for development, because it lets us push distinct
images tags and avoid confusion when deploying new manifests on a test
environment.

However, it will take more storage on quay.io.

### CODE REVIEW

**Comment by VannTen** [comment]
That's a proposal, and there is some drawbacks.

We would also needs to tag the repo with tags.

**Comment by goern** [comment]
having tags is normal mode of development for us, isnt it? so where is the drawback? ;)

**Comment by VannTen** [comment]
More images tags (basically, one each time each of us needs to test a version and push the image tag to quay.io) so more storage.
I'm pretty sold on that (ofc, wouldn't propose otherwise) but I feel obligated to emphasize potential drawbacks.

**Comment by codificat** [comment]
I am not completely sold on this (yet).

I think that for self-testing you can always use a separate registry location. The moment you need to share something, even if draft, you would likely submit a PR - and at that point, we can already create an image for that, right? E.g. I believe that this should create an image based on this current PR:

/deploy

Then, for already-merged changes, I think we should rely on the actual tags. If a (merged) change is worth a new "release" (it could be minor, or just a patch) then we should also signal that by creating the corresponding tag for it, no?  And at that point, we get an image for the newly created tag...

So, in other words: it seems to me that for the work happening on the repo we should rely on PRs (for WIP changes) or tags (for merged changes) only, and use other "personal" repositories for changes that are not yet ready to share.

File: Makefile [hunk 1]
```diff
@@ -4,7 +4,7 @@
 # - use the VERSION as arg of the bundle target (e.g make bundle VERSION=0.0.2)
 # - use environment variables to overwrite this value (e.g export VERSION=0.0.2)
 
-_VERSION := $(shell git describe --tags --dirty)
+_VERSION := $(shell git describe --tags --dirty --always)
```
Comment (codificat) [inline Makefile:7]: I like the idea of syncing the version to the tags.

But if we stick to "release tags" for "accepted" versions I would instead do:

```suggestion
_VERSION := $(shell git describe --tags)
```

as there ShouldNot™ be an image without its mathing repo tag - but if there is, you would still get the long format of `git describe` with num commits and hash

**Comment by codificat** [comment]
> /deploy

well, that didn't have the reaction I was expecting... no image was built for the PR :disappointed:

**Comment by VannTen** [comment]
> I am not completely sold on this (yet).
> 
> I think that for self-testing you can always use a separate registry location. The moment you need to share something, even if draft, you would likely submit a PR - and at that point, we can already create an image for that, right? E.g. I believe that this should create an image based on this current PR:

I just go and idea for doing both. The IMG var could be:
```makefile
IMG := $(REGISTRY_URL)/$(NAMESPACE)/$(IMAGE):$(VERSION)
```
With NAMESPACE defaulting to $USER (or somehting) like that and mapping to a personnal image namespace on quay.io (or somewhere else). Meanwhile the CI on PR would set those vars to a shared values (thoth-station... ) and publish here. Wdyt ? 

> Then, for already-merged changes, I think we should rely on the actual tags. If a (merged) change is worth a new "release" (it could be minor, or just a patch) then we should also signal that by creating the corresponding tag for it, no? And at that point, we get an image for the newly created tag...

Absolutely. I was mainly thinking of the dev side while making the PR. The --always and --dirty flags are debatable.

One thing, though, do we have a working deployment per PR right now ?

**Comment by codificat** [comment]
> With NAMESPACE defaulting to $USER (or somehting) like that and mapping to a personnal image namespace on quay.io (or somewhere else). Meanwhile the CI on PR would set those vars to a shared values (thoth-station... ) and publish here. Wdyt ?

We would still have to be setting env vars... and for that, we can already set `IMAGE_TAG_BASE` (and `VERSION`) as needed. I might be missing something, but I don't see what would be the benefit of adding these.

> Absolutely. I was mainly thinking of the dev side while making the PR. The --always and --dirty flags are debatable.

Right. In my opinion, they don't belong in the `main` branch - and for feature branches, we can rely on:
 
> One thing, though, do we have a working deployment per PR right now ?

Yes, but just need to:
- use `/deploy` as a single standalone comment [as in the comment above](https://github.com/thoth-station/meteor-operator/pull/114#issuecomment-1300233474) (or change the [current code](https://github.com/AICoE/aicoe-ci/blob/775b31e00794ddf3d68e1673313474d2fb9c2a10/conditions/deploy-comment-check.yaml#L25) handling this)
- adjust the base image in the `Dockerfile` so that it doesn't use a short name, or somehow fix the build process for the current configuration

The image build pipeline triggered by `/deploy` above failed with:

```
+ buildah bud --storage-driver=overlay --tls-verify=false --layers -f /gen-source/Containerfile -t meteor-operator-114 .
[1/2] STEP 1/10: FROM golang:1.18 AS builder
error creating build container: short-name resolution enforced but cannot prompt without a TTY
```

**Comment by VannTen** [comment]
On Wed, Nov 02, 2022 at 06:29:39AM -0700, Pep Turró Mauri wrote:
> > With NAMESPACE defaulting to $USER (or somehting) like that and mapping to a personnal image namespace on quay.io (or somewhere else). Meanwhile the CI on PR would set those vars to a shared values (thoth-station... ) and publish here. Wdyt ?
> 
> We would still have to be setting env vars... and for that, we can already set `IMAGE_TAG_BASE` (and `VERSION`) as needed. I might be missing something, but I don't see what would be the benefit of adding these.

I was going at this with a "no-setup" angle for the dev side (aka: make "deploy" builds everything, push, apply the manifest etc). But without using a flat shared namespace, it's not very pratical, for sure ; I don't see any local identifier which could be mapped to a personal registry namespace easily.
 
> - use `/deploy` as a single standalone comment [as in the comment above](https://github.com/thoth-station/meteor-operator/pull/114#issuecomment-1300233474) (or change the [current code](https://github.com/AICoE/aicoe-ci/blob/775b31e00794ddf3d68e1673313474d2fb9c2a10/conditions/deploy-comment-check.yaml#L25) handling this)

This builds the image but does not apply the manifests anywhere, right ?
And the resulting tag is predictable ? (So we can push my branch, apply
my manifests on my quicklab instance (or similar) without waiting for
the build to finish)
